### PR TITLE
feat: add install and uninstall sub cmd to egctl

### LINKED ISSUE #2711: feat: add `egctl install` and `egctl uninstall`
*Description*:

Have we considered adding more installation methods for **eg**?

Currently we support installation through `k8s resource file` and `helm`. 

In my opinion, we can add `egctl install` command line method for installation and `egctl uninstall` method for uninstallation.

This can reduce other issues that users need to pay attention to, and reduces the installation steps compared to the other two methods. This is more friendly to users who are less familiar with k8s/helm.

### PR DESCRIPTION
**Pull request by ShyunnY**
**What type of PR is this?**

feat: Add install and uninstall subcommands to egctl

**Which issue(s) this PR fixes**:

Fixes #2711

### CODE REVIEW

**Comment by arkodg** [comment]
@alexwo meet @ShyunnY , @ShyunnY meet @alexwo :)
looks like you both are trying to create an internal helm lib to install and uninstall EG https://github.com/envoyproxy/gateway/pull/2859
would be great if the lib can be reused between `egctl` and `e2e`

**Comment by ShyunnY** [comment]
> @alexwo meet @ShyunnY , @ShyunnY meet @alexwo :) looks like you both are trying to create an internal helm lib to install and uninstall EG #2859 would be great if the lib can be reused between `egctl` and `e2e`

This looks good and I think a more general option could be made using the HelmTool I wrote. :)

**Comment by ShyunnY** [comment]
@arkodg Hello! I migrated HelmTool to the `/utils/helm` package and renamed it to PackageTool. Consider this naming because it bears the package management responsibilities of `eg` (e.g.  installation, uninstallation, and upgrades that may be added in the future, etc.). The use of the Helm library may not be limited to `egctl`. As you said, I pulled it out as a tool that others can use.

**Comment by arkodg** [comment]
cc @alexwo

**Comment by zirain** [comment]
should we put these under `experimetal` first?

**Comment by ShyunnY** [comment]
This is a good idea.

File: internal/utils/helm/package.go
Comment (shawnh2) [inline internal/utils/helm/package.go:240]: typo `failed`

File: internal/utils/helm/package.go
Comment (shawnh2) [inline internal/utils/helm/package.go:315]: this is a fixed value now, imo users should be able to install EG under any namespace, this can be beneficial to EG's [multi-tenancy deployment mode](https://gateway.envoyproxy.io/v1.0.0/user/operations/deployment-mode/#multi-tenancy).

Comment (zirain) [reply internal/utils/helm/package.go:315]: should be configurable via `--namspace` (`-n` in short)

File: internal/cmd/egctl/experimental.go
Comment (shawnh2) [inline internal/cmd/egctl/experimental.go:29]: hi @ShyunnY, the code looks good! 

since we have decided to set these two subcmds as experimental, can we add a section `egctl experimental install & uninstall` in https://gateway.envoyproxy.io/latest/user/operations/egctl/ to describe how to use these cmds.

and also, since we can install EG using egctl, maybe this doc https://gateway.envoyproxy.io/latest/install/install-egctl/ can be renamed as `install-with-egctl`, and how to install egctl will be a section in it. WDYT?

Comment (ShyunnY) [reply internal/cmd/egctl/experimental.go:29]: IMO, I think https://gateway.envoyproxy.io/latest/install/install-egctl/ does not need to be changed, this part of the doc is meant to tell us how to install egctl, we are not supposed to add other information here. Currently install/uninstall is only in experimental status, so it is enough for us to just add description information at https://gateway.envoyproxy.io/latest/user/operations/egctl/.

When we later consider changing install/uninstall to stable, we will consider adding additional docs to specifically describe how to use egctl to install or uninstall eg.

I think this is more appropriate, what do you think? :)

Comment (shawnh2) [reply internal/cmd/egctl/experimental.go:29]: SGTM

**Review by shawnh2 — APPROVED** [review]
LGTM! Also test against the [multi-tenancy mode](https://gateway.envoyproxy.io/v1.0.0/user/operations/deployment-mode/#multi-tenancy) and it works as expected. Thanks @ShyunnY 

There're some conflicts, can you resolve?

**Comment by ShyunnY** [comment]
I may not be able to resolve this conflict at the moment, I may need help

**Comment by ShyunnY** [comment]
@arkodg
It seems that I can't solve the go.mod problem. If I modify it according to lint, conflicts will occur. If I modify the conflicts, it will cause lint problems.

Also, do we really need to add tests?

**Comment by arkodg** [comment]
sorry about the go mod rebase issues, will make sure to not merge any more go.mod PRs

**Comment by arkodg** [comment]
instead of a test, lets raise a follow up to leverage this tool lib in our e2e

**Comment by ShyunnY** [comment]
> instead of a test, lets raise a follow up to leverage this tool lib in our e2e

Since I couldn't resolve the `go.mod` conflict issue, I chose to undo the changes to `go.mod`, which may require troublesome troubleshooting after the merge, thank you again! 

After waiting for the code to be merged, I will follow up on the writing of related e2e.

File: internal/utils/helm/package.go
Comment (arkodg) [inline internal/utils/helm/package.go:134]:
```suggestion
	installCmd.Flags().BoolVar(&opts.SkipCRD, "skip-crds", false, "if set, no CRDs will be installed. By default, CRDs are installed if not already present")
```

File: internal/utils/helm/package.go
Comment (arkodg) [inline internal/utils/helm/package.go:135]:
```suggestion
	installCmd.Flags().BoolVar(&opts.OnlyCRD, "only-crds", false, "if set, only install the crd")
```

File: internal/utils/helm/package.go
Comment (arkodg) [inline internal/utils/helm/package.go:151]:
```suggestion
	uninstallCmd.Flags().BoolVar(&opts.WithCRD, "with-crds", false, "if set, the CRDs will also be removed")
```

File: internal/utils/helm/package.go
Comment (arkodg) [inline internal/utils/helm/package.go:131]: maybe just `name` ?

**Comment by arkodg** [comment]
minor comments around naming, thanks for building this out !

**Comment by ShyunnY** [comment]
> minor comments around naming, thanks for building this out !

Thanks for your review! I will fix the test issue as soon as possible, thank you again!

File: internal/utils/helm/package.go
Comment (ShyunnY) [reply internal/utils/helm/package.go:131]: > maybe just `name` ?

I have thought about this problem, but I think it is more suitable to use `release-name`.

+ it's self-explanatory
+ release-name does not seem to be easily ambiguous, at least when compared with name.

Comment (arkodg) [reply internal/utils/helm/package.go:131]: we ideally like typing less when its comes to CLI commands, and the flag description does mention that this is related to release.
Lets get this in, and get feedback, since its experimental, we can always improve this later

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Pull request merged**
